Add ES6 and React JavaScript styleguides

curioussavage opened the pull request:
The old javascript guide was also moved to /es5

:eyeglasses: @ajacksified @schwers @nramadas @madbook @dwick and whoever else cares

Updated with the latest version of the Airbnb style guide and applied the same handful of changes that we made to the original. I also added their React styleguide and removed a few things we don't do.

Getting a little weary of scanning around this so seems like a good time for input.

Here are some specific parts to look at.

JS
--------
* Arrays
  use slice to copy array too?

* Arrow functions
  what styles do we prefer?

* comments
  fixme /todo convention?

* Naming conventions
  possible rule; name constants with all caps?

React
----------
* ordering (do we even care that much? conventions might be nice though)

schwers commented on javascript/react/README.md:
```diff
@@ -0,0 +1,325 @@
+
+## Basic Rules
+
+  - Only include one React component per file.
```
I don't think anyone's commented on this, but: https://github.com/schwers/reddit-mobile/blob/topnav/src/views/components/OverlayMenuRow.jsx

Context: In topnav I'm adding a component to draw overlays, and there's a file dedicated to specific kinds of rows. Right now the row's are relatively simple so I kept them in the same file so they could share come common code without dealing with subclasses. I'm not sure if this is an example of where this rule breaks down, or if I should just change that file. Open to suggestions, thoughts?

curioussavage replied:
I also have a small subcomponent in the same file so I would lean towards not being so strict. If a component is really only meant to be used with another then I think it makes sense not to be so strict

dwick replied:
:+1: for subcomponents only used by the parent.

madbook replied:
:+1:

ajacksified replied:
:+1:

schwers commented on javascript/react/README.md:
```diff
@@ -0,0 +1,325 @@
+
+## Class vs `React.createClass`
+
+  - Use `class extends React.Component` unless you have a very good reason to use mixins.
```
It seems like the React teams wants to remove mixins so I think we should avoid them completely.

madbook replied:
:+1: to avoiding mixins altogether

ajacksified replied:
:+1:

schwers commented on javascript/react/README.md:
````diff
@@ -0,0 +1,325 @@
+
+    ```javascript
+    // bad
+    <Foo superLongParam="bar"
````
These examples should all be single quotes

curioussavage replied:
yeah, a lot of the examples have stuff that does not meet our style :frowning:

madbook replied:
:no_bell: as a personal preference, i actually still like to use double-quotes in jsx.  I think it makes it a little easier to distinguish the jsx junk from the js junk, but it's mostly force-of-habit

ajacksified replied:
I prefer single quotes everywhere; we shouldn't have js strings in JSX that aren't template strings using ` ` ` anyways.

schwers replied:
@ajacksified what about strings like "You don't like double quotes?". I think I'm starting to lean towards double quotes

schwers commented on javascript/react/README.md:
````diff
@@ -0,0 +1,325 @@
+
+    // good
+    <Foo />
+    ```
````
A section on brackets and the space rules for them would be good

curioussavage replied:
brackets or curly braces?

schwers replied:
(whoops) curly braces

ajacksified replied:
our linter likes `{ spaces }`

schwers commented on javascript/react/README.md:
```diff
@@ -0,0 +1,325 @@
+
+  - Bind event handlers for the render method in the constructor.
+
+  > Why? A bind call in a the render path creates a brand new function on every single render.
```
Might want to mention our opinion of things like
 ```
class Bad {
  poorlyBoundMethod = () => {
  }
}

vs

class Good {
  constructor() {
    this.properlyBoundMethod = this.properlyBoundMethod.bind(this);
  }
  properlyBoundMethod() {
  }
}
```

madbook replied:
:no_bell: what is the reason for preferring the latter form here?

schwers replied:
Babel outputs less than ideal code in the Bad case. Specifically it creates a new copy of the function `poorlyBoundMethod` everytime the constructor runs as opposed to binding the function from the prototype.

```
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Bad = function Bad() {
  _classCallCheck(this, Bad);

  this.poorlyBoundMethod = function () {};
};

vs;

var Good = (function () {
  function Good() {
    _classCallCheck(this, Good);

    this.properlyBoundMethod = this.properlyBoundMethod.bind(this);
  }

  _createClass(Good, [{
    key: "properlyBoundMethod",
    value: function properlyBoundMethod() {}
  }]);

  return Good;
})();
```

schwers commented on javascript/react/README.md:
```diff
@@ -0,0 +1,325 @@
+
+## Ordering
+
+  - Ordering for `class extends React.Component`:
```
:no_bell: I like this idea of ordering. Though after talking about it with @nramadas I think the first or second should be `static propTypes {...}`. Thoughts?

dwick replied:
i think static stuff should go before the constructor, including `propTypes`

ajacksified replied:
I prefer it at the bottom, since in 90% of cases I skip it over to get to the logic anyways... but I just want it to be consistent, so I don't feel strongly about where.

ajacksified replied:
Edit: talked in person, let's go :arrow_up: to the top

schwers commented:
Can you add a section on splats?

dwick commented:
does react have support for linters? would be nice to have some configuration files with this.

:nail_care:

florenceyeun commented on javascript/react/README.md:
```diff
@@ -0,0 +1,325 @@
+# Airbnb React/JSX Style Guide
```
`Airbnb` --> `Reddit`?

schwers commented:
[Couldn't find a way to comment on this because the default pull view isn't showing the diff for README.md and es5/README.md -- PSA check those out if you haven't]

It would be good to come up with a standard for curly brackets. The rule `object-curly-spacing` in README.md says no spaces inside, but we use spaces in `jsx` templates and `setState` calls. Having spaces inside only for `jsx` is basically what we do now so I'd vote to always have spaces for consistency's sake.

schwers commented:
Should have a rule for whitespace for function declarations. eg `function() {}` versus `function () {}`. We mostly use the space, except functions in classes are inconsistent right now. I think we should just always have the space for consistency and easier grep'ing

schwers commented:
:nail_care:

curioussavage commented:
@dwick we have an eslint config file in our project with ES6 and React specific rules. We should probably store it here too

madbook commented on javascript/react/README.md:
```diff
@@ -0,0 +1,325 @@
+    // good
+    <Foo
+      userName="hello"
+      phoneNumber={12345678}
```
this example has no spaces inside the curly braces, which is defined as being `bad` in the next section O_O

curioussavage replied:
hah, yeah. I will need to go through the styleguide and make sure it passes its own rules. Just waiting until I stop making changes. no linter to help find issues here :(

ajacksified commented on javascript/react/README.md:
```diff
@@ -0,0 +1,325 @@
+
+  - Use `class extends React.Component` unless you have a very good reason to use mixins.
+
+  eslint rules: [`react/prefer-es6-class`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prefer-es6-class.md).
```
How about a rule about preferring stateless components?

madbook replied:
^ :+1:

ajacksified commented on javascript/react/README.md:
```diff
@@ -0,0 +1,325 @@
+## Naming
+
+  - **Extensions**: Use `.jsx` extension for React components.
+  - **Filename**: Use PascalCase for filenames. E.g., `ReservationCard.jsx`.
```
I think we should _always_ go lowercase, so that we don't accidentally screw up imports on case-sensitive filesystems, which has happened several times in the past.

madbook replied:
^ can confirm :+1:

schwers replied:
:+1:

dwick replied:
:+1: :+1: :+1:

curioussavage replied:
yeah this one actually bit me the other day so :+1: :+1:

ajacksified commented on javascript/react/README.md:
```diff
@@ -0,0 +1,325 @@
+
+    // good
+    <Foo
+      hidden
```
:no_bell: I dislike this, personally; I like explicitness.

ajacksified commented on javascript/react/README.md:
```diff
@@ -0,0 +1,325 @@
+
+    // good, when single line
+    render() {
+      const body = <div>hello</div>;
```
How do you feel about wrapping all react in parens? We've been doing that pretty much everywhere. (I think it's more readable.)

curioussavage replied:
I think that is a good idea

dwick replied:
:+1:

ajacksified commented:
https://github.com/curioussavage/styleguide/tree/update-js-rules/javascript#functions - @schwers has used `const functionName = function functionName () { }`, which, while verbose, explicitly calls out a function as a `const` instead of a `var`.

ajacksified commented:
The "good" example in https://github.com/curioussavage/styleguide/tree/update-js-rules/javascript#8.1 uses parens around `x`, which should be removed per rule #8.2.

ajacksified commented:
I'd remove the "ok" example from [#10.1](https://github.com/curioussavage/styleguide/tree/update-js-rules/javascript#10.1)

curioussavage commented:
:haircut:  ok everybody I think I addressed all the comments. Maybe everyone just look at the issues the brought up first to check since there are so many.

curioussavage commented:
I did not add anything assigning a named function to a const, any opinions?

curioussavage commented:
going to add a rule barring people from using expressions like this for control flow:
`country = canGeotarget && $('#country').val() || '',`

schwers commented:
@curioussavage from our talk with @nramadas that example should probably use a ternary instead of a boolean expression.

schwers commented:
:fish: :tropical_fish: :party:

ajacksified commented:
:tada:
